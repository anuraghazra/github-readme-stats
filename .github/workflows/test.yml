name: Test

on:
  push:
    branches:
      - "*"
  pull_request:
    branches:
      - master

jobs:
  build:
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v2

      - name: Setup Node
        uses: actions/setup-node@v1
        with:
          node-version: "12.x"

      - name: Cache node modules
        uses: actions/cache@v2
        env:
          cache-name: cache-node-modules
        with:
          path: ~/.npm
          key:
            ${{ runner.os }}-npm-cache-${{ hashFiles('**/package-lock.json') }}
          restore-keys: |
            ${{ runner.os }}-npm-cache-

      - name: Install & Test
        run: |
          npm install
          npm run test

      - name: Code Coverage
        uses: codecov/codecov-action@v1
