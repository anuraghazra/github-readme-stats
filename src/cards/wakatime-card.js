const Card = require("../common/Card");
const I18n = require("../common/I18n");
const { getStyles } = require("../getStyles");
const { wakatimeCardLocales } = require("../translations");
const languageColors = require("../common/languageColors.json");
const { createProgressNode } = require("../common/createProgressNode");
const {
  clampValue,
  getCardColors,
  flexLayout,
  lowercaseTrim,
} = require("../common/utils");

const noCodingActivityNode = ({ color, text }) => {
  return `
    <text x="25" y="11" class="stat bold" fill="${color}">${text}</text>
  `;
};

const createCompactLangNode = ({ lang, totalSize, x, y }) => {
  const color = languageColors[lang.name] || "#858585";

  return `
    <g transform="translate(${x}, ${y})">
      <circle cx="5" cy="6" r="5" fill="${color}" />
      <text data-testid="lang-name" x="15" y="10" class='lang-name'>
        ${lang.name} - ${lang.text}
      </text>
    </g>
  `;
};

const createLanguageTextNode = ({ langs, totalSize, x, y }) => {
  return langs.map((lang, index) => {
    if (index % 2 === 0) {
      return createCompactLangNode({
        lang,
        x: 25,
        y: 12.5 * index + y,
        totalSize,
        index,
      });
    }
    return createCompactLangNode({
      lang,
      x: 230,
      y: 12.5 + 12.5 * index,
      totalSize,
      index,
    });
  });
};

const createTextNode = ({
  id,
  label,
  value,
  index,
  percent,
  hideProgress,
  progressBarColor,
  progressBarBackgroundColor,
}) => {
  const staggerDelay = (index + 3) * 150;

  const cardProgress = hideProgress
    ? null
    : createProgressNode({
        x: 110,
        y: 4,
        progress: percent,
        color: progressBarColor,
        width: 220,
        name: label,
        progressBarBackgroundColor,
      });

  return `
    <g class="stagger" style="animation-delay: ${staggerDelay}ms" transform="translate(25, 0)">
      <text class="stat bold" y="12.5" data-testid="${id}">${label}:</text>
      <text
        class="stat"
        x="${hideProgress ? 170 : 350}"
        y="12.5"
      >${value}</text>
      ${cardProgress}
    </g>
  `;
};

const recalculatePercentages = (languages) => {
  // recalculating percentages so that,
  // compact layout's progress bar does not break when hiding languages
  const totalSum = languages.reduce(
    (totalSum, language) => totalSum + language.percent,
    0,
  );
  const weight = (100 / totalSum).toFixed(2);
  languages.forEach((language) => {
    language.percent = (language.percent * weight).toFixed(2);
  });
};

const renderWakatimeCard = (stats = {}, options = { hide: [] }) => {
  let { languages } = stats;
  const {
    hide_title = false,
    hide_border = false,
    hide,
    line_height = 25,
    title_color,
    icon_color,
    text_color,
    bg_color,
    theme = "default",
    hide_progress,
    custom_title,
    locale,
    layout,
    langs_count = languages ? languages.length : 0,
    border_radius,
    border_color,
  } = options;

  const shouldHideLangs = Array.isArray(hide) && hide.length > 0;
  if (shouldHideLangs) {
    const languagesToHide = new Set(hide.map((lang) => lowercaseTrim(lang)));
    languages = languages.filter(
      (lang) => !languagesToHide.has(lowercaseTrim(lang.name)),
    );
    recalculatePercentages(languages);
  }

  const i18n = new I18n({
    locale,
    translations: wakatimeCardLocales,
  });

  const lheight = parseInt(line_height, 10);

  langsCount = clampValue(parseInt(langs_count), 1, langs_count);

  // returns theme based colors with proper overrides and defaults
  const {
    titleColor,
    textColor,
    iconColor,
    bgColor,
    borderColor,
  } = getCardColors({
    title_color,
    icon_color,
    text_color,
    bg_color,
    border_color,
    theme,
  });

  const filteredLanguages = languages
    ? languages
        .filter((language) => language.hours || language.minutes)
        .slice(0, langsCount)
    : [];

  // Calculate the card height depending on how many items there are
  // but if rank circle is visible clamp the minimum height to `150`
  let height = Math.max(45 + (filteredLanguages.length + 1) * lheight, 150);

  const cssStyles = getStyles({
    titleColor,
    textColor,
    iconColor,
  });

  let finalLayout = "";

  let width = 440;

  // RENDER COMPACT LAYOUT
  if (layout === "compact") {
    width = width + 50;
    height = 90 + Math.round(filteredLanguages.length / 2) * 25;

    // progressOffset holds the previous language's width and used to offset the next language
    // so that we can stack them one after another, like this: [--][----][---]
    let progressOffset = 0;
    const compactProgressBar = filteredLanguages
      .map((language) => {
        // const progress = (width * lang.percent) / 100;
        const progress = ((width - 25) * language.percent) / 100;

        const languageColor = languageColors[language.name] || "#858585";

        const output = `
          <rect
            mask="url(#rect-mask)"
            data-testid="lang-progress"
            x="${progressOffset}"
            y="0"
            width="${progress}"
            height="8"
            fill="${languageColor}"
          />
        `;
        progressOffset += progress;
        return output;
      })
      .join("");

    finalLayout = `
      <mask id="rect-mask">
      <rect x="25" y="0" width="${width - 50}" height="8" fill="white" rx="5" />
      </mask>
      ${compactProgressBar}
      ${createLanguageTextNode({
        x: 0,
        y: 25,
        langs: filteredLanguages,
        totalSize: 100,
      }).join("")}
    `;
  } else {
    finalLayout = flexLayout({
      items: filteredLanguages.length
        ? filteredLanguages.map((language) => {
            return createTextNode({
              id: language.name,
              label: language.name,
              value: language.text,
              percent: language.percent,
              progressBarColor: titleColor,
              progressBarBackgroundColor: textColor,
              hideProgress: hide_progress,
            });
          })
        : [
            noCodingActivityNode({
              color: textColor,
              text: i18n.t("wakatimecard.nocodingactivity"),
            }),
          ],
      gap: lheight,
      direction: "column",
    }).join("");
  }

  const card = new Card({
    customTitle: custom_title,
    defaultTitle: i18n.t("wakatimecard.title"),
    width: 495,
    height,
    border_radius,
    colors: {
      titleColor,
      textColor,
      iconColor,
      bgColor,
      borderColor,
    },
  });

  card.setHideBorder(hide_border);
  card.setHideTitle(hide_title);
  card.setCSS(
    `
    ${cssStyles}
    .lang-name { font: 400 11px 'Segoe UI', Ubuntu, Sans-Serif; fill: ${textColor} }
    `,
  );

  return card.render(`
    <svg x="0" y="0" width="100%">
      ${finalLayout}
    </svg>
  `);
};

module.exports = renderWakatimeCard;
exports.createProgressNode = createProgressNode;
