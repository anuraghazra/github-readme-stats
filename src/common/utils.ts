import axios from "axios";
import { themes } from "../../themes";
import wrap from "word-wrap";

export function renderError(message: string, secondaryMessage = ""): string {
  return `
    <svg width="495" height="120" viewBox="0 0 495 120" fill="none" xmlns="http://www.w3.org/2000/svg">
    <style>
    .text { font: 600 16px 'Segoe UI', Ubuntu, Sans-Serif; fill: #2F80ED }
    .small { font: 600 12px 'Segoe UI', Ubuntu, Sans-Serif; fill: #252525 }
    .gray { fill: #858585 }
    </style>
    <rect x="0.5" y="0.5" width="494" height="99%" rx="4.5" fill="#FFFEFE" stroke="#E4E2E2"/>
    <text x="25" y="45" class="text">Something went wrong! file an issue at https://git.io/JJmN9</text>
    <text data-testid="message" x="25" y="55" class="text small">
      <tspan x="25" dy="18">${encodeHTML(message)}</tspan>
      <tspan x="25" dy="18" class="gray">${secondaryMessage}</tspan>
    </text>
    </svg>
  `;
}

//? https://stackoverflow.com/a/48073476/10629172
export function encodeHTML(str: string): string {
  return str
    .replace(/[\u00A0-\u9999<>&](?!#)/gim, (i) => {
      return "&#" + i.charCodeAt(0) + ";";
    })
    .replace(/\u0008/gim, "");
}

export function kFormatter(num: number): string {
  return Math.abs(num) > 999
    ? Math.sign(num) * parseInt((Math.abs(num) / 1000).toFixed(1)) + "k"
    : (Math.sign(num) * Math.abs(num)).toString();
}

export function isValidHexColor(hexColor: string) {
  return new RegExp(
    /^([A-Fa-f0-9]{8}|[A-Fa-f0-9]{6}|[A-Fa-f0-9]{3}|[A-Fa-f0-9]{4})$/,
  ).test(hexColor);
}

/**
 * Returns false if the string did not match "true"
 * @param value string to parse
 */
export function parseBoolean(value: string): boolean {
  return value === "true" ? true : false;
}

export function parseArray(str: string): string[] {
  return str ? str.split(",") : [];
}

export function clampValue(number: number, min: number, max: number): number {
  return Math.max(min, Math.min(number, max));
}

function isValidGradient(colors: string[]): boolean {
  return isValidHexColor(colors[1]) && isValidHexColor(colors[2]);
}

export function fallbackColor(color: string, fallbackColor: string) {
  const colors = color.split(",");
  let gradient: string[] = null;

  if (colors.length > 1 && isValidGradient(colors)) {
    gradient = colors;
  }

  return (
    (gradient ? gradient : isValidHexColor(color) && `#${color}`) ||
    fallbackColor
  );
}

export function request(data, headers) {
  return axios({
    url: "https://api.github.com/graphql",
    method: "post",
    headers,
    data,
  });
}

/**
 * Auto layout utility, allows us to layout things
 * vertically or horizontally with proper gaping
 */
export function FlexLayout({
  items,
  gap,
  direction,
}: {
  items: string[];
  gap: number;
  direction?: string;
}) {
  //* filter() for filtering out empty strings
  return items.filter(Boolean).map((item, i) => {
    let transform = `translate(${gap * i}, 0)`;
    if (direction === "column") {
      transform = `translate(0, ${gap * i})`;
    }
    return `<g transform="${transform}">${item}</g>`;
  });
}

//* returns theme based colors with proper overrides and defaults
export function getCardColors({
  title_color,
  text_color,
  icon_color,
  bg_color,
  theme,
  fallbackTheme = "default",
}: {
  title_color?: string;
  text_color?: string;
  icon_color?: string;
  bg_color?: string;
  theme?: string;
  fallbackTheme?: string;
}) {
  const defaultTheme = themes[fallbackTheme],
    selectedTheme = themes[theme] || defaultTheme,
    //* get the color provided by the user else the theme color
    //* finally if both colors are invalid fallback to default theme
    titleColor = fallbackColor(
      title_color || selectedTheme.title_color,
      "#" + defaultTheme.title_color,
    ),
    iconColor = fallbackColor(
      icon_color || selectedTheme.icon_color,
      "#" + defaultTheme.icon_color,
    ),
    textColor = fallbackColor(
      text_color || selectedTheme.text_color,
      "#" + defaultTheme.text_color,
    ),
    bgColor = fallbackColor(
      bg_color || selectedTheme.bg_color,
      "#" + defaultTheme.bg_color,
    );

  return { titleColor, iconColor, textColor, bgColor };
}

export function wrapTextMultiline(text: string, width = 60, maxLines = 3) {
  const wrapped = wrap(encodeHTML(text), { width })
    .split("\n") //* Split wrapped lines to get an array of lines
    .map((line) => line.trim()); //* Remove leading and trailing whitespace of each line

  const lines = wrapped.slice(0, maxLines); //* Only consider maxLines lines

  //* Add "..." to the last line if the text exceeds maxLines
  if (wrapped.length > maxLines) {
    lines[maxLines - 1] += "...";
  }

  //* Remove empty lines if text fits in less than maxLines lines
  const multiLineText = lines.filter(Boolean);
  return multiLineText;
}

const noop = () => {};
//* return console instance based on the environment
export const logger =
  process.env.NODE_ENV !== "test" ? console : { log: noop, error: noop };

export const CONSTANTS = {
  THIRTY_MINUTES: 1800,
  TWO_HOURS: 7200,
  FOUR_HOURS: 14400,
  ONE_DAY: 86400,
};

const SECONDARY_ERROR_MESSAGES = {
  MAX_RETRY:
    "Please add an env variable called PAT_1 with your github token in vercel",
  USER_NOT_FOUND: "Make sure the provided username is not an organization",
};

export class CustomError extends Error {
  type: string;
  secondaryMessage: string;

  constructor(message, type) {
    super(message);
    this.type = type;
    this.secondaryMessage = SECONDARY_ERROR_MESSAGES[type] || "adsad";
  }

  static MAX_RETRY = "MAX_RETRY";
  static USER_NOT_FOUND = "USER_NOT_FOUND";
}

//? https://stackoverflow.com/a/48172630/10629172
export function measureText(str: string, fontSize = 10): number {
  // prettier-ignore
  const widths = [
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0.2796875, 0.2765625,
    0.3546875, 0.5546875, 0.5546875, 0.8890625, 0.665625, 0.190625,
    0.3328125, 0.3328125, 0.3890625, 0.5828125, 0.2765625, 0.3328125,
    0.2765625, 0.3015625, 0.5546875, 0.5546875, 0.5546875, 0.5546875,
    0.5546875, 0.5546875, 0.5546875, 0.5546875, 0.5546875, 0.5546875,
    0.2765625, 0.2765625, 0.584375, 0.5828125, 0.584375, 0.5546875,
    1.0140625, 0.665625, 0.665625, 0.721875, 0.721875, 0.665625,
    0.609375, 0.7765625, 0.721875, 0.2765625, 0.5, 0.665625,
    0.5546875, 0.8328125, 0.721875, 0.7765625, 0.665625, 0.7765625,
    0.721875, 0.665625, 0.609375, 0.721875, 0.665625, 0.94375,
    0.665625, 0.665625, 0.609375, 0.2765625, 0.3546875, 0.2765625,
    0.4765625, 0.5546875, 0.3328125, 0.5546875, 0.5546875, 0.5,
    0.5546875, 0.5546875, 0.2765625, 0.5546875, 0.5546875, 0.221875,
    0.240625, 0.5, 0.221875, 0.8328125, 0.5546875, 0.5546875,
    0.5546875, 0.5546875, 0.3328125, 0.5, 0.2765625, 0.5546875,
    0.5, 0.721875, 0.5, 0.5, 0.5, 0.3546875, 0.259375, 0.353125, 0.5890625,
  ];

  const avg = 0.5279276315789471;
  return (
    str
      .split("")
      .map((c) =>
        c.charCodeAt(0) < widths.length ? widths[c.charCodeAt(0)] : avg,
      )
      .reduce((cur, acc) => acc + cur) * fontSize
  );
}
