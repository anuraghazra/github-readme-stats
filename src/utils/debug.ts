export const logger =
  process.env.NODE_ENV !== "test"
    ? console
    : { log: (...args: any[]) => void 0, error: (...args: any[]) => void 0 };
