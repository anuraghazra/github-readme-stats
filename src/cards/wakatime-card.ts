import { FlexLayout, getCardColors } from "../common/utils";

import { Card } from "../common/Card";
import { I18n } from "../common/I18n";
import { createProgressNode } from "../common/createProgressNode";
import { getStyles } from "../getStyles";
import languageColors from "../common/languageColors.json";
import { wakatimeCardLocales } from "../translations";

function noCodingActivityNode({
  color,
  text,
}: {
  color: string | string[];
  text: string;
}) {
  return `
    <text x="25" y="11" class="stat bold" fill="${color}">${text}</text>
  `;
}

function createCompactLangNode({
  lang,
  x,
  y,
}: {
  lang: { name: string; text: string };
  x: number;
  y: number;
}) {
  const color = languageColors[lang.name] || "#858585";

  return `
    <g transform="translate(${x}, ${y})">
      <circle cx="5" cy="6" r="5" fill="${color}" />
      <text data-testid="lang-name" x="15" y="10" class='lang-name'>
        ${lang.name} - ${lang.text}
      </text>
    </g>
  `;
}

function createLanguageTextNode({
  langs,
  y,
}: {
  langs: {
    name: string;
    text: string;
  }[];
  y: number;
}) {
  return langs.map((lang, index: number) => {
    if (index % 2 === 0) {
      return createCompactLangNode({
        lang,
        x: 25,
        y: 12.5 * index + y,
      });
    }
    return createCompactLangNode({
      lang,
      x: 230,
      y: 12.5 + 12.5 * index,
    });
  });
}

function createTextNode({
  id,
  label,
  value,
  index,
  percent,
  hideProgress,
  progressBarColor,
  progressBarBackgroundColor,
}) {
  const staggerDelay = (index + 3) * 150,
    cardProgress = hideProgress
      ? null
      : createProgressNode({
          x: 110,
          y: 4,
          progress: percent,
          color: progressBarColor,
          width: 220,
          progressBarBackgroundColor,
        });

  return `
    <g class="stagger" style="animation-delay: ${staggerDelay}ms" transform="translate(25, 0)">
      <text class="stat bold" y="12.5">${label}:</text>
      <text
        class="stat"
        x="${hideProgress ? 170 : 350}"
        y="12.5"
        data-testid="${id}"
      >${value}</text>
      ${cardProgress}
    </g>
  `;
}

export function renderWakatimeCard(
  stats: {
    languages?: {
      hours?: number;
      minutes?: number;
      name: string;
      text: string;
      percent;
    }[];
  } = {},
  options: {
    hide_title?: boolean;
    hide_border?: boolean;
    line_height?: number;
    title_color?: string;
    icon_color?: string;
    text_color?: string;
    bg_color?: string;
    theme?: string;
    hide_progress?: boolean;
    custom_title?: string;
    locale?: string;
    layout?: string;
  } = {},
) {
  const { languages } = stats;
  const {
    hide_title = false,
    hide_border = false,
    line_height = 25,
    title_color,
    icon_color,
    text_color,
    bg_color,
    theme = "default",
    hide_progress,
    custom_title,
    locale,
    layout,
  } = options;

  const i18n = new I18n({
      locale,
      translations: wakatimeCardLocales,
    }),
    lheight = parseInt(line_height.toString(), 10),
    //* returns theme based colors with proper overrides and defaults
    { titleColor, textColor, iconColor, bgColor } = getCardColors({
      title_color,
      icon_color,
      text_color,
      bg_color,
      theme,
    }),
    statItems = languages
      ? languages
          .filter((language) => language.hours || language.minutes)
          .map((language, i) => {
            return createTextNode({
              id: language.name,
              label: language.name,
              value: language.text,
              index: i,
              percent: language.percent,
              progressBarColor: titleColor,
              progressBarBackgroundColor: textColor,
              hideProgress: hide_progress,
            });
          })
      : [];

  //* Calculate the card height depending on how many items there are
  //* but if rank circle is visible clamp the minimum height to `150`
  let height = Math.max(45 + (statItems.length + 1) * lheight, 150),
    finalLayout = "",
    width = 440;

  const cssStyles = getStyles({
    titleColor,
    textColor,
    iconColor,
  });

  //! RENDER COMPACT LAYOUT
  if (layout === "compact") {
    width = width + 50;
    height = 90 + Math.round(languages.length / 2) * 25;

    //* progressOffset holds the previous language's width and used to offset the next language
    //* so that we can stack them one after another, like this: [--][----][---]
    let progressOffset = 0;
    const compactProgressBar = languages
      .map((lang) => {
        //? const progress = (width * lang.percent) / 100;
        const progress = ((width - 50) * lang.percent) / 100,
          languageColor = languageColors[lang.name] || "#858585",
          output = `
          <rect
            mask="url(#rect-mask)"
            data-testid="lang-progress"
            x="${progressOffset}"
            y="0"
            width="${progress}"
            height="8"
            fill="${languageColor}"
          />
        `;
        progressOffset += progress;
        return output;
      })
      .join("");

    finalLayout = `
      <mask id="rect-mask">
      <rect x="25" y="0" width="${width - 50}" height="8" fill="white" rx="5" />
      </mask>
      ${compactProgressBar}
      ${createLanguageTextNode({
        y: 25,
        langs: languages,
      }).join("")}
    `;
  } else {
    finalLayout = FlexLayout({
      items: statItems.length
        ? statItems
        : [
            noCodingActivityNode({
              color: textColor,
              text: i18n.t("wakatimecard.nocodingactivity"),
            }),
          ],
      gap: lheight,
      direction: "column",
    }).join("");
  }

  const card = new Card({
    customTitle: custom_title,
    defaultTitle: i18n.t("wakatimecard.title"),
    width: 495,
    height,
    colors: {
      titleColor,
      textColor,
      iconColor,
      bgColor,
    },
  });

  card.setHideBorder(hide_border);
  card.setHideTitle(hide_title);
  card.setCSS(
    `
    ${cssStyles}
    .lang-name { font: 400 11px 'Segoe UI', Ubuntu, Sans-Serif; fill: ${textColor} }
    `,
  );

  return card.render(`
    <svg x="0" y="0" width="100%">
      ${finalLayout}
    </svg>
  `);
}
