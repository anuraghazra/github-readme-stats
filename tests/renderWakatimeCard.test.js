import { queryByTestId } from "@testing-library/dom";
import "@testing-library/jest-dom";
import { renderWakatimeCard } from "../src/cards/wakatime-card.js";
import { getCardColors } from "../src/common/utils.js";
import { wakaTimeData } from "./fetchWakatime.test.js";

describe("Test Render Wakatime Card", () => {
  it("should render correctly", () => {
    const card = renderWakatimeCard(wakaTimeData.data);
    expect(getCardColors).toMatchSnapshot();
  });

  it("should render correctly with compact layout", () => {
    const card = renderWakatimeCard(wakaTimeData.data, { layout: "compact" });

    expect(card).toMatchSnapshot();
  });

  it("should render correctly with compact layout when langs_count is set", () => {
    const card = renderWakatimeCard(wakaTimeData.data, {
      layout: "compact",
      langs_count: 2,
    });

    expect(card).toMatchSnapshot();
  });

  it("should hide languages when hide is passed", () => {
    document.body.innerHTML = renderWakatimeCard(wakaTimeData.data, {
      hide: ["YAML", "Other"],
    });

    expect(queryByTestId(document.body, /YAML/i)).toBeNull();
    expect(queryByTestId(document.body, /Other/i)).toBeNull();
    expect(queryByTestId(document.body, /TypeScript/i)).not.toBeNull();
  });

  it("should render translations", () => {
    document.body.innerHTML = renderWakatimeCard({}, { locale: "cn" });
    expect(document.getElementsByClassName("header")[0].textContent).toBe(
      "Wakatime 周统计",
    );
    expect(
      document.querySelector('g[transform="translate(0, 0)"]>text.stat.bold')
        .textContent,
    ).toBe("本周没有编程活动");
  });

  it("should render without rounding", () => {
    document.body.innerHTML = renderWakatimeCard(wakaTimeData.data, {
      border_radius: "0",
    });
    expect(document.querySelector("rect")).toHaveAttribute("rx", "0");
    document.body.innerHTML = renderWakatimeCard(wakaTimeData.data, {});
    expect(document.querySelector("rect")).toHaveAttribute("rx", "4.5");
  });

  it('should show "no coding activitiy this week" message when there hasn not been activity', () => {
    document.body.innerHTML = renderWakatimeCard(
      {
        ...wakaTimeData.data,
        languages: undefined,
      },
      {},
    );
    expect(document.querySelector(".stat").textContent).toBe(
      "No coding activity this week",
    );
  });
});
