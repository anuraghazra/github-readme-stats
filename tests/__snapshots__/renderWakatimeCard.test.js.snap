// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Render Wakatime Card should render correctly 1`] = `
"
      <svg
        width=\\"495\\"
        height=\\"150\\"
        viewBox=\\"0 0 495 150\\"
        fill=\\"none\\"
        xmlns=\\"http://www.w3.org/2000/svg\\"
        role=\\"img\\"
        aria-labelledby=\\"descId\\"
      >
        <title id=\\"titleId\\"></title>
        <desc id=\\"descId\\"></desc>
        <style>
          .header {
            font: 600 18px 'Segoe UI', Ubuntu, Sans-Serif;
            fill: #2f80ed;
            animation: fadeInAnimation 0.8s ease-in-out forwards;
          }
          @supports(-moz-appearance: auto) {
            /* Selector detects Firefox */
            .header { font-size: 15.5px; }
          }
          
    
    .stat {
      font: 600 14px 'Segoe UI', Ubuntu, \\"Helvetica Neue\\", Sans-Serif; fill: #434d58;
    }
    @supports(-moz-appearance: auto) {
      /* Selector detects Firefox */
      .stat { font-size:12px; }
    }
    .stagger {
      opacity: 0;
      animation: fadeInAnimation 0.3s ease-in-out forwards;
    }
    .rank-text {
      font: 800 24px 'Segoe UI', Ubuntu, Sans-Serif; fill: #434d58; 
      animation: scaleInAnimation 0.3s ease-in-out forwards;
    }
    
    .bold { font-weight: 700 }
    .icon {
      fill: #4c71f2;
      display: none;
    }
    
    .rank-circle-rim {
      stroke: #2f80ed;
      fill: none;
      stroke-width: 6;
      opacity: 0.2;
    }
    .rank-circle {
      stroke: #2f80ed;
      stroke-dasharray: 250;
      fill: none;
      stroke-width: 6;
      stroke-linecap: round;
      opacity: 0.8;
      transform-origin: -10px 8px;
      transform: rotate(-90deg);
      animation: rankAnimation 1s forwards ease-in-out;
    }
    
  
    .lang-name { font: 400 11px 'Segoe UI', Ubuntu, Sans-Serif; fill: #434d58 }
    

          
          
        </style>

        

        <rect
          data-testid=\\"card-bg\\"
          x=\\"0.5\\"
          y=\\"0.5\\"
          rx=\\"4.5\\"
          height=\\"99%\\"
          stroke=\\"#e4e2e2\\"
          width=\\"494\\"
          fill=\\"#fffefe\\"
          stroke-opacity=\\"1\\"
        />

        
      <g
        data-testid=\\"card-title\\"
        transform=\\"translate(25, 35)\\"
      >
        <g transform=\\"translate(0, 0)\\">
      <text
        x=\\"0\\"
        y=\\"0\\"
        class=\\"header\\"
        data-testid=\\"header\\"
      >Wakatime Stats</text>
    </g>
      </g>
    

        <g
          data-testid=\\"main-card-body\\"
          transform=\\"translate(0, 55)\\"
        >
          
    <svg x=\\"0\\" y=\\"0\\" width=\\"100%\\">
      <g transform=\\"translate(0, 0)\\">
    <g class=\\"stagger\\" style=\\"animation-delay: NaNms\\" transform=\\"translate(25, 0)\\">
      <text class=\\"stat bold\\" y=\\"12.5\\" data-testid=\\"Other\\">Other:</text>
      <text
        class=\\"stat\\"
        x=\\"350\\"
        y=\\"12.5\\"
      >19 mins</text>
      
    <svg width=\\"220\\" x=\\"110\\" y=\\"4\\">
      <rect rx=\\"5\\" ry=\\"5\\" x=\\"0\\" y=\\"0\\" width=\\"220\\" height=\\"8\\" fill=\\"#434d58\\"></rect>
      <rect
          height=\\"8\\"
          fill=\\"#2f80ed\\"
          rx=\\"5\\" ry=\\"5\\" x=\\"0\\" y=\\"0\\" 
          data-testid=\\"lang-progress\\"
          width=\\"2%\\"
      >
      </rect>
    </svg>
  
    </g>
  </g><g transform=\\"translate(0, 25)\\">
    <g class=\\"stagger\\" style=\\"animation-delay: NaNms\\" transform=\\"translate(25, 0)\\">
      <text class=\\"stat bold\\" y=\\"12.5\\" data-testid=\\"TypeScript\\">TypeScript:</text>
      <text
        class=\\"stat\\"
        x=\\"350\\"
        y=\\"12.5\\"
      >1 min</text>
      
    <svg width=\\"220\\" x=\\"110\\" y=\\"4\\">
      <rect rx=\\"5\\" ry=\\"5\\" x=\\"0\\" y=\\"0\\" width=\\"220\\" height=\\"8\\" fill=\\"#434d58\\"></rect>
      <rect
          height=\\"8\\"
          fill=\\"#2f80ed\\"
          rx=\\"5\\" ry=\\"5\\" x=\\"0\\" y=\\"0\\" 
          data-testid=\\"lang-progress\\"
          width=\\"2%\\"
      >
      </rect>
    </svg>
  
    </g>
  </g>
    </svg>
  
        </g>
      </svg>
    "
`;

exports[`Test Render Wakatime Card should render correctly with compact layout 1`] = `
"
      <svg
        width=\\"495\\"
        height=\\"115\\"
        viewBox=\\"0 0 495 115\\"
        fill=\\"none\\"
        xmlns=\\"http://www.w3.org/2000/svg\\"
        role=\\"img\\"
        aria-labelledby=\\"descId\\"
      >
        <title id=\\"titleId\\"></title>
        <desc id=\\"descId\\"></desc>
        <style>
          .header {
            font: 600 18px 'Segoe UI', Ubuntu, Sans-Serif;
            fill: #2f80ed;
            animation: fadeInAnimation 0.8s ease-in-out forwards;
          }
          @supports(-moz-appearance: auto) {
            /* Selector detects Firefox */
            .header { font-size: 15.5px; }
          }
          
    
    .stat {
      font: 600 14px 'Segoe UI', Ubuntu, \\"Helvetica Neue\\", Sans-Serif; fill: #434d58;
    }
    @supports(-moz-appearance: auto) {
      /* Selector detects Firefox */
      .stat { font-size:12px; }
    }
    .stagger {
      opacity: 0;
      animation: fadeInAnimation 0.3s ease-in-out forwards;
    }
    .rank-text {
      font: 800 24px 'Segoe UI', Ubuntu, Sans-Serif; fill: #434d58; 
      animation: scaleInAnimation 0.3s ease-in-out forwards;
    }
    
    .bold { font-weight: 700 }
    .icon {
      fill: #4c71f2;
      display: none;
    }
    
    .rank-circle-rim {
      stroke: #2f80ed;
      fill: none;
      stroke-width: 6;
      opacity: 0.2;
    }
    .rank-circle {
      stroke: #2f80ed;
      stroke-dasharray: 250;
      fill: none;
      stroke-width: 6;
      stroke-linecap: round;
      opacity: 0.8;
      transform-origin: -10px 8px;
      transform: rotate(-90deg);
      animation: rankAnimation 1s forwards ease-in-out;
    }
    
  
    .lang-name { font: 400 11px 'Segoe UI', Ubuntu, Sans-Serif; fill: #434d58 }
    

          
          
        </style>

        

        <rect
          data-testid=\\"card-bg\\"
          x=\\"0.5\\"
          y=\\"0.5\\"
          rx=\\"4.5\\"
          height=\\"99%\\"
          stroke=\\"#e4e2e2\\"
          width=\\"494\\"
          fill=\\"#fffefe\\"
          stroke-opacity=\\"1\\"
        />

        
      <g
        data-testid=\\"card-title\\"
        transform=\\"translate(25, 35)\\"
      >
        <g transform=\\"translate(0, 0)\\">
      <text
        x=\\"0\\"
        y=\\"0\\"
        class=\\"header\\"
        data-testid=\\"header\\"
      >Wakatime Stats</text>
    </g>
      </g>
    

        <g
          data-testid=\\"main-card-body\\"
          transform=\\"translate(0, 55)\\"
        >
          
    <svg x=\\"0\\" y=\\"0\\" width=\\"100%\\">
      
      <mask id=\\"rect-mask\\">
      <rect x=\\"25\\" y=\\"0\\" width=\\"440\\" height=\\"8\\" fill=\\"white\\" rx=\\"5\\" />
      </mask>
      
          <rect
            mask=\\"url(#rect-mask)\\"
            data-testid=\\"lang-progress\\"
            x=\\"0\\"
            y=\\"0\\"
            width=\\"6.6495\\"
            height=\\"8\\"
            fill=\\"#858585\\"
          />
        
          <rect
            mask=\\"url(#rect-mask)\\"
            data-testid=\\"lang-progress\\"
            x=\\"6.6495\\"
            y=\\"0\\"
            width=\\"0.465\\"
            height=\\"8\\"
            fill=\\"#2b7489\\"
          />
        
      
    <g transform=\\"translate(25, 25)\\">
      <circle cx=\\"5\\" cy=\\"6\\" r=\\"5\\" fill=\\"#858585\\" />
      <text data-testid=\\"lang-name\\" x=\\"15\\" y=\\"10\\" class='lang-name'>
        Other - 19 mins
      </text>
    </g>
  
    <g transform=\\"translate(230, 25)\\">
      <circle cx=\\"5\\" cy=\\"6\\" r=\\"5\\" fill=\\"#2b7489\\" />
      <text data-testid=\\"lang-name\\" x=\\"15\\" y=\\"10\\" class='lang-name'>
        TypeScript - 1 min
      </text>
    </g>
  
    
    </svg>
  
        </g>
      </svg>
    "
`;
